File: packages/nativescript/platforms/ios/NitroPushNativeScript.swift
import Foundation

@objc(NitroPushNativeScript)
public final class NitroPushNativeScript: NSObject {
    @objc public static let shared = NitroPushNativeScript()
    private let queue = DispatchQueue(label: "org.nitropush.nativescript")
    private var configured = false

    /// Called by the generated main before allocating NativeScript/V8.
    @objc public static func selectRoot(_ embeddedRoot: String) -> String {
        let bridge = shared
        do {
            func required(_ name: String) throws -> String {
                guard let value = Bundle.main.object(forInfoDictionaryKey: name) as? String, !value.isEmpty else {
                    throw NSError(domain: "NitroPush", code: 1, userInfo: nil)
                }
                return value
            }
            let runtime = try required("NITROPUSH_RUNTIME_VERSION")
            guard runtime.range(of: "^[a-f0-9]{64}$", options: .regularExpression) != nil else { return embeddedRoot }
            try NitroPushSdk.shared.configure(NPConfig(
                deploymentKey: try required("NITROPUSH_DEPLOYMENT_KEY"),
                appVersion: runtime,
                bundlePublicKey: try required("NITROPUSH_BUNDLE_PUBLIC_KEY")
            ))
            bridge.configured = true
            let root = NitroPushSdk.shared.verifiedApplicationRoot(embeddedRoot: URL(fileURLWithPath: embeddedRoot))
            if let root = root { return root.path }
            if NitroPushSdk.shared.getCurrentPackage() != nil {
                // No unverified selection may be reported healthy by the embedded app.
                NitroPushSdk.shared.clearUpdates()
            }
        } catch { bridge.configured = false }
        return embeddedRoot
    }

    @objc public func execute(_ operation: String, completion: @escaping (String?, String?) -> Void) {
        queue.async {
            guard self.configured else {
                DispatchQueue.main.async { completion(nil, "NitroPush bootstrap/configuration is missing; rebuild the native app") }
                return
            }
            func done(_ result: String?, _ error: String? = nil) {
                DispatchQueue.main.async { completion(result, error) }
            }
            switch operation {
            case "ready": NitroPushSdk.shared.notifyAppReady(); done("null")
            case "current": done(self.json(NitroPushSdk.shared.getCurrentPackage()))
            case "pending": done(self.json(NitroPushSdk.shared.getPendingPackage()))
            case "clearPending": NitroPushSdk.shared.clearPendingUpdate(); done("null")
            case "sync":
                let semaphore = DispatchSemaphore(value: 0)
                Task {
                    defer { semaphore.signal() }
                    do {
                        if NitroPushSdk.shared.getPendingPackage() != nil {
                            done("{\"status\":\"UPDATE_INSTALLED\"}"); return
                        }
                        guard let remote = try await NitroPushSdk.shared.checkForUpdate(deploymentKeyOverride: nil) else {
                            done("{\"status\":\"UP_TO_DATE\"}"); return
                        }
                        let local = try await NitroPushSdk.shared.downloadUpdate(remote)
                        try await NitroPushSdk.shared.installUpdate(pkg: local, installMode: .onNextRestart, minimumBackgroundDuration: 0)
                        done("{\"status\":\"UPDATE_INSTALLED\"}")
                    } catch { done(nil, "NitroPush update failed verification or download") }
                }
                semaphore.wait()
            default: done(nil, "Unsupported NitroPush operation")
            }
        }
    }

    private func json(_ package: NPLocalPackage?) -> String {
        guard let package = package else { return "null" }
        let value: [String: Any] = ["releaseId": package.releaseId, "label": package.label,
            "appVersion": package.appVersion, "isPending": package.isPending]
        guard let data = try? JSONSerialization.data(withJSONObject: value) else { return "null" }
        return String(data: data, encoding: .utf8) ?? "null"
    }
}

File: packages/nativescript/platforms/android/java/com/nitropush/sdk/NitroPushNativeScript.kt
package com.nitropush.sdk

import android.content.Context
import android.content.pm.PackageManager
import android.os.Handler
import android.os.Looper
import java.io.File
import java.util.concurrent.Executors
import org.json.JSONObject

class NitroPushNativeScript private constructor() {
    private val executor = Executors.newSingleThreadExecutor()
    private val main = Handler(Looper.getMainLooper())
    private var configured = false
    @Volatile private var applicationRoot = ""

    companion object {
        private val instance = NitroPushNativeScript()
        @JvmStatic fun getInstance() = instance

        /** Injected after APK extraction, before AppConfig and Runtime initialization. */
        @JvmStatic fun selectRoot(context: Context, embeddedRoot: File): File {
            instance.applicationRoot = File(embeddedRoot, "app").absolutePath
            if ((context.applicationInfo.flags and android.content.pm.ApplicationInfo.FLAG_DEBUGGABLE) != 0) return embeddedRoot
            return runCatching {
                val metadata = context.packageManager.getApplicationInfo(context.packageName, PackageManager.GET_META_DATA).metaData
                fun required(name: String) = metadata?.getString(name)?.takeIf { it.isNotBlank() } ?: error("Missing native configuration")
                val runtime = required("NITROPUSH_RUNTIME_VERSION")
                check(Regex("^[a-f0-9]{64}$").matches(runtime))
                NitroPushSdk.install(context)
                NitroPushSdk.shared.configure(NPConfig(
                    deploymentKey = required("NITROPUSH_DEPLOYMENT_KEY"),
                    appVersion = runtime,
                    bundlePublicKey = required("NITROPUSH_BUNDLE_PUBLIC_KEY"),
                ))
                instance.configured = true
                val root = NitroPushSdk.shared.verifiedApplicationRoot(embeddedRoot)
                if (root == null) {
                    if (NitroPushSdk.shared.getCurrentPackage() != null) NitroPushSdk.shared.clearUpdates()
                    embeddedRoot
                } else {
                    // These runtime-owned files always come from this APK, never from OTA.
                    for (name in listOf("internal", "metadata")) {
                        val source = File(embeddedRoot, name)
                        if (source.exists()) check(source.copyRecursively(File(root, name), overwrite = true))
                    }
                    instance.applicationRoot = File(root, "app").absolutePath
                    root
                }
            }.getOrElse { instance.configured = false; embeddedRoot }
        }
    }

    /** NativeScript's JS file APIs otherwise keep resolving ~/ against filesDir/app. */
    fun getApplicationRoot(): String = applicationRoot

    fun execute(operation: String, callback: NitroPushCallback) {
        executor.execute {
            var result: String? = null
            var error: String? = null
            var phase = "check"
            try {
                check(configured) { "NitroPush bootstrap/configuration is missing" }
                val core = NitroPushSdk.shared
                result = when (operation) {
                    "ready" -> { core.notifyAppReady(); "null" }
                    "current" -> json(core.getCurrentPackage())
                    "pending" -> json(core.getPendingPackage())
                    "clearPending" -> { core.clearPendingUpdate(); "null" }
                    "sync" -> {
                        if (core.getPendingPackage() != null) "{\"status\":\"UPDATE_INSTALLED\"}"
                        else {
                            val remote = core.checkForUpdate()
                            if (remote == null) "{\"status\":\"UP_TO_DATE\"}"
                            else {
                                phase = "download/verification"
                                val local = core.downloadUpdate(remote)
                                phase = "install"
                                core.installUpdate(local, NPInstallMode.ON_NEXT_RESTART, 0.0)
                                "{\"status\":\"UPDATE_INSTALLED\"}"
                            }
                        }
                    }
                    else -> error("Unsupported NitroPush operation")
                }
            } catch (_: Throwable) {
                error = if (!configured) "NitroPush bootstrap/configuration is missing; rebuild the native app"
                else "Update $phase failed. Check the deployment key, connection, signing public key and runtime version."
            }
            main.post { callback.complete(result, error) }
        }
    }

    private fun json(pkg: NPLocalPackage?): String = if (pkg == null) "null" else JSONObject()
        .put("releaseId", pkg.releaseId).put("label", pkg.label)
        .put("appVersion", pkg.appVersion).put("isPending", pkg.isPending).toString()
}
